Optimize simple prefix and suffix LIKE expressions

kaikalur opened the pull request:
Cleaned up the LIKE prefix implementation and also added suffix versions.

Test plan - added tests

```
== NO RELEASE NOTE ==
```

feilong-liu commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -7048,4 +7048,17 @@ private static boolean assertSorted(List<Object> array)
+    public void testLikePrefixAndSuffix()
+    {
+        assertQuery("select x like 'abc%' from (values 'abc', 'def', 'bcd') T(x)");
+        assertQuery("select x like '%abc%' from (values 'xabcy', 'abxabcdef', 'bcd') T(x)");
```
Also add 'xabcyabcz' which has multiplle 'abc'?

kaikalur replied:
done

feilong-liu commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -7048,4 +7048,17 @@ private static boolean assertSorted(List<Object> array)
+    {
+        assertQuery("select x like 'abc%' from (values 'abc', 'def', 'bcd') T(x)");
+        assertQuery("select x like '%abc%' from (values 'xabcy', 'abxabcdef', 'bcd') T(x)");
+        assertQuery("select x like '%abc' from (values 'xa bc', 'xabcy', 'abcd') T(x)");
```
Also add a match of 'xabc'

kaikalur replied:
done

pranjalssh commented on presto-main/src/main/java/com/facebook/presto/sql/relational/SqlToRowExpressionTranslator.java:
```diff
@@ -913,28 +919,44 @@ protected RowExpression visitLikePredicate(LikePredicate node, Context context)
-                    if (patternString.length() > 1 && !patternString.contains("_")) {
-                        if (patternString.indexOf('%') == patternString.length() - 1) {
+                    int matchLength = patternString.length();
+                    if (matchLength > 1 && !patternString.contains("_")) {
```
Negate this if and return early. Less indentation is good!

kaikalur replied:
> Negate this if and return early. Less indentation is good!

Single return from methods is better!

tdcmeehan commented:
CC: @kaikalur

tdcmeehan commented:
@psantos-denodo thanks for flagging this.  This is definitely a bug, and I think we should put out a patch fix for this in 0.285.  I believe the best thing to do is to just restrict this optimization for `VARCHAR`, as otherwise there's complications around blank padding semantics if we overload `split` to take in a char and return a char.

https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/relational/SqlToRowExpressionTranslator.java#L932

We should remove the check on CharType, as it's simply not supported:

` || value.getType() instanceof CharType`

And add a test.

kaikalur commented:
> @kaikalur if you can provide a fix, we can patch it for 0.284 and 0.285, and make sure it lands for 0.286

Actually the code is still the same! Surprised we haven't caught that sooner - almost 1.5yrs. Is char type widely used (in hive connector) or is it just a jdbc one?

tdcmeehan commented:
The use of char vs. varchar is highly dependent on the connectors used/one's data infrastructure and the conventions of the data engineers exposing the data.  It's very hard to say how widely it's used, but it is fair to say that one particular deployment may have come across this sooner than others.

It's been present since 0.284, and considering many users lag several versions behind, it's doesn't surprise me we catch this now.

kaikalur commented:
> Hi, I'm experiencing an error executing a query in Presto 0.285 that was working in version Presto 0.283. Based on the error traces it looks like it can be due to this feature. The query is: `select * from customer t2 WHERE t2.c_first_name LIKE '%Daisy%'` where customer is in a JDBC source and c_first_name has type char(20)
> 
> In that case the query fails with message: `SQL Error [6]: Query failed (#20240228_181216_00060_hpywp): Unexpected parameters (char(20), varchar, bigint) for function split. Expected: split(varchar(x), varchar(y), bigint) , split(varchar(x), varchar(y)) `
> 
> It only fails if I use % at the begging and the end ('%Daisy%'). It works with '%Daisy' and 'Daisy%' . It works if I use a varchar column instead of char(20) but char(20) works in older versions.

Can we also check if the other cases in that code fail - say just sufix or just prefix - LIKE '%a', LIKE 'a%'. Wonder if the issue is with SPLIT and maybe fix that for a general case.

Put another way,

Can you try:  

SPLIT(t2.c_first_name, 'Daisy')

And also

substr(t2.c_first_name, 1, length('Daisy'))

I want to understand what the issue is. 

also can you open a github issue instead of comments on a closed PR? Thanks!

tdcmeehan commented:
@kaikalur went ahead and created #22045
